src/util/userDataListUtil.js

```diff
@@ -0,0 +1,33 @@
+function GetClientObjects(userDataList) {
+  let clientObjectList = [];
+
+  userDataList.forEach(
+    ({ name: userName, client: userClient, vendor: userVendor }) => {
+      // find or create client object
+      let clientObject = clientObjectList.find(
+        ({ name: clientName }) => clientName === userClient
+      );
+      if (clientObject == null) {
+        clientObject = { name: userClient, vendors: [] };
+        clientObjectList.push(clientObject);
+      }
+
+      // find or create vendor object
+      let vendorObject = clientObject.vendors.find(
+        ({ name: vendorName }) => vendorName === userVendor
+      );
+      if (vendorObject == null) {
+        vendorObject = { name: userVendor, users: [] };
+        clientObject.vendors.push(vendorObject);
+      }
+
+      // add user
+      let userObject = { name: userName };
+      vendorObject.users.push(userObject);
+    }
+  );
```

Comment:
There are some fancy ways to do this kind of thing using an awkward combination of reduce and some other things, but I prefer this style tbh haha. I'm not too sure how they do it in Endeavour, as we don't really do a ton of data manipulation. The most complex thing I did was this:
```
export const buildEmailAddresses = (
  addresses: EmailOptionsAddresses,
  vendorSupportedLocales: LocaleDefinition[],
) =>
  vendorSupportedLocales.reduce(
    (newAddresses: any, { iso639: isoCode }: LocaleDefinition) => {
      const name = addresses?.[isoCode]?.name ?? '';
      const address = addresses?.[isoCode]?.address ?? '';
      return {
        ...newAddresses,
        [isoCode]: {
          name: name,
          address: address,
          useDefaultAddress: name === '' && address === '',
        },
      };
    },
    {},
  );
```